### DISCUSSION TITLE
Keycloak 18 as windows service

### DISCUSSION THREAD

**Discussion by Turbo-Pascal** [keycloak/keycloak#12879] [Q&A]
Is there a recommended way to run keycloak 18 as a windows service? 

Since there is no longer a service.bat in v18 and kc.bat, kcadm.bat and kcreg.bat don't seem to have any options for running as a service, is there still an option to run keycloak 18 as a windows service?

**Comment by nullfx** [comment]
@Turbo-Pascal did you ever get a response to this? We're about to re-engage a major client who we are about to roll through the process of updating their infrastructure (they're on v13) and rolling them up to the latest version but from my experience then and now it doesn't appear that this product has any sort of customization or *actual* documentation anywhere around.  it seems like they have 1 way that the system can be deployed (and as far as I can tell having run it that way, it doesn't work very well).

looks like we'll be migrating our customer over to cognito unless somebody has any sort of *actual* success stories about being able to use a current version of keycloak or deploy it either in the cloud or (preferrably) as a windows service on prem.

- very frustrated user.

**Reply by vmuzikar** [reply]
@nullfx Hello, sorry to hear you are facing issues when using Keycloak on Win platform. I'd suggest creating a GH Issue with steps to reproduce. That would help us in a great way with triaging and eventually fixing that issue. Thank you.

**Reply by Turbo-Pascal** [reply]
@nullfx We abandoned the windows hosting option. Since quarkus it seems to be not on the dev team's objectives: no docs, no reference projects or success stories here. We compared several products but in the end we wanted to stick with keycloak. So we 'followed the flow' of the docs and switched to a container environment and there it performs really well.

**Comment by Xulunix** [comment]
I'm running 20.0.5 as windows service using NSSM and a batch file to set environment variables and start keycloak.
Works without issues for me so far.

**Reply by viniprince39** [reply]
Hi can you give an example on how to create a batch file to start the keycloak as service.?
i created the services easily in Linux but not able to create in Windows. can you please help?

**Reply by Xulunix** [reply]
As i said, i used NSSM to host keycloak as a service. No need for a batch file.
https://nssm.cc/

**Comment by brmfsantos** [comment]
Good afternoon,

I'm setting up Keycloak in a Windows environment.

I've already installed Keycloak, ran the command to start it, and it installed the database.

When I run NSSM to create the service, I get the service created, but when I try to start it, it always remains as "Paused."

I don't see any errors in the log viewer.

Does anyone know what might be happening?

[enter image description here](https://i.sstatic.net/CgcyRjrk.png)

I created the service and when I tried to start it, it was expected to be working, but it fails to start and sets its status to "paused."

**Comment by brmfsantos** [comment]
Good afternoon,

I'm setting up Keycloak in a Windows environment.

I've already installed Keycloak, ran the command to start it, and it installed the database.

When I run NSSM to create the service, I get the service created, but when I try to start it, it always remains as "Paused."

I don't see any errors in the log viewer.

Does anyone know what might be happening?

[enter image description here](https://i.sstatic.net/CgcyRjrk.png)

I created the service and when I tried to start it, it was expected to be working, but it fails to start and sets its status to "paused."

**Reply by brmfsantos** [reply]
My Solution:

Can use the following commands:
nssm install Keycloak "D:\Keycloak\keycloak-26.0.5\bin\kc.bat"
nssm set "Keycloak" AppDirectory "D:\Keycloak\keycloak-26.0.5\bin"
nssm set "Keycloak" AppParameters start
nssm set "Keycloak" Description "Keycloak Identity and Access Management"
